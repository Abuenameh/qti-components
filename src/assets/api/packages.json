{
  "packages": [
    "biologie",
    "replay",
    "toetsenincontext",
    "kringloop"
  ]
}
