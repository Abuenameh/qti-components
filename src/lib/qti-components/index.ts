/* only exported to override or extend for specific purposes */
export * from './internal/expression-result';
export * from './internal/event-types';
export * from './internal/variables';
export * from './internal/events';
/* only exported to override or extend for specific purposes */
export * from './qti-assessment-item/qti-assessment-item';
export * from './qti-stylesheet/qti-stylesheet';
export * from './qti-item-body/qti-item-body';
export * from './qti-prompt/qti-prompt';
export * from './qti-interaction/internal/choice/qti-choice';

/* start response and outcome */
export * from './qti-variable-declaration/qti-response-declaration/qti-response-declaration';
export * from './qti-variable-declaration/qti-outcome-declaration/qti-outcomedeclaration';
/* end response and outcome */

export * from './qti-companion-materials-info/qti-companion-materials-info';
export * from './qti-rubric-block/qti-rubric-block';
export * from './qti-rubric-block/qti-content-body';

/* start feedback */
export * from './qti-feedback/qti-feedback-inline/qti-feedback-inline';
export * from './qti-feedback/qti-modal-feedback/qti-modal-feedback';
export * from './qti-feedback/qti-feedback-block/qti-feedback-block';

/* end feedback */

/* start textentryinteraction */
export * from './qti-interaction/qti-text-entry-interaction/qti-text-entry-interaction';
export * from './qti-interaction/qti-extended-text-interaction/qti-extended-text-interaction';

/* end textentryinteraction */

/* start hottext */
export * from './qti-interaction/qti-hottext-interaction/qti-hottext-interaction';
/* end hottext */

/* start inlinechoice */
export * from './qti-interaction/qti-inline-choice-interaction/qti-inline-choice-interaction';
/* end inlinechoice */

/* start choiceinteraction */
export * from './qti-interaction/qti-choice-interaction/qti-choice-interaction';
/* end choiceinteraction */

export * from './qti-response-processing';

/* start custom interactions */
export * from './qti-interaction/qti-portable-custom-interaction/qti-portable-custom-interaction';
/* end custom interactions */

/* start only preview items */
export * from './qti-interaction/qti-associate-interaction/qti-associate-interaction';

export * from './qti-interaction/qti-gap-match-interaction/qti-gap-match-interaction';

export * from './qti-interaction/qti-graphic-associate-interaction/qti-graphic-associate-interaction';

export * from './qti-interaction/qti-graphic-gap-match-interaction/qti-graphic-gap-match-interaction';

export * from './qti-interaction/qti-graphic-order-interaction/qti-graphic-order-interaction';

export * from './qti-interaction/qti-hotspot-interaction/qti-hotspot-interaction';

export * from './qti-interaction/qti-match-interaction/qti-match-interaction';

export * from './qti-interaction/qti-media-interaction/qti-media-interaction';

export * from './qti-interaction/qti-order-interaction/qti-order-interaction';

export * from './qti-interaction/qti-position-object-interaction/qti-position-object-interaction';
export * from './qti-interaction/qti-position-object-interaction/qti-position-object-stage';

export * from './qti-interaction/qti-select-point-interaction/qti-select-point-interaction';

export * from './qti-interaction/qti-slider-interaction/qti-slider-interaction';

export * from './qti-interaction/qti-end-attempt-interaction/qti-end-attempt-interaction';

/* end only preview items */

export * from './qti-interaction/qti-associable-hotspot';
export * from './qti-interaction/qti-gap-img';
export * from './qti-interaction/qti-gap-text';
export * from './qti-interaction/qti-gap';
export * from './qti-interaction/qti-hotspot-choice';
export * from './qti-interaction/qti-hottext';
export * from './qti-interaction/qti-inline-choice';
export * from './qti-interaction/qti-simple-associable-choice';
export * from './qti-interaction/qti-simple-choice';
export * from './qti-interaction/internal/interaction/interaction';

export * from './qti-assessment-item/qti-assessment-item.context';
