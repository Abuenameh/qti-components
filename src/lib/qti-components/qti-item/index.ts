export * from '../qti-assessment-item/qti-item.context';
export * from '../qti-assessment-item/qti-item.context.controller';
export * from './qti-item';
export * from './qti-item-logger';
