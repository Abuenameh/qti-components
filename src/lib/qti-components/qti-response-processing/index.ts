export { default as QtiResponseProcessing } from './qti-response-processing/qti-responseprocessing';
export { QtiRule } from './qti-rule/qti-rule';

export * from './qti-rule/qti-response-condition/qti-response-condition';
export * from './qti-rule/qti-set-outcome-value/qti-set-outcome-value';

export * from './qti-response-else/qti-response-if/qti-response-if';
export * from './qti-response-else/qti-response-if/qti-response-else-if/qti-response-else-if';
export * from './qti-response-else/qti-response-else';

export * from './qti-expression';
