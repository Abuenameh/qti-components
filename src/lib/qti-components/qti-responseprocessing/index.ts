export { default as QtiResponseProcessing } from './qti-responseprocessing/qti-responseprocessing';
export { QtiRule } from './qti-rule/qti-rule';

export * from './qti-rule/qti-responsecondition/qti-responsecondition';
export * from './qti-rule/qti-setoutcomevalue/qti-setoutcomevalue';

export * from './qti-responseelse/qti-responseif/qti-responseif';
export * from './qti-responseelse/qti-responseif/qti-responseelseif/qti-responseelseif';
export * from './qti-responseelse/qti-responseelse';

export * from './qti-expression';
