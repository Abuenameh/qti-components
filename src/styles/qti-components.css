@layer base {
  /** default styles for all components */
  .base {
    @apply cursor-pointer rounded-md 
      outline-none ring-1 ring-inset ring-gray-300;
  }
}

@layer components {
  /** The different components, you can choose only one */

  /* - button  : .input-control-hidden qti-choice, qti-hotspot */
  .button {
    @apply base px-3 py-2 font-semibold shadow-sm;
    /* @apply select-none; */
  }

  /*  - select  : inline-choice */
  .select {
    @apply base relative appearance-none  px-3 py-2 pr-7 shadow-sm;
    background: url("data:image/svg+xml,%3Csvg fill='currentColor' width='22' viewBox='0 0 20 20' xmlns='http://www.w3.org/2000/svg' aria-hidden='true'%3E%3Cpath clip-rule='evenodd' fill-rule='evenodd' d='M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z'%3E%3C/path%3E%3C/svg%3E")
      no-repeat center right 6px;
  }

  /* - check   : qti-choice, qti-hotspot */
  .check {
    @apply base flex items-center gap-2 rounded-full px-1 py-0.5 shadow-none;
    @apply !ring-0; /* !important, remove for checkbox and radio the active ring, the radio and checkbox will be selected instead */
    /* @apply select-none; */

    &::part(cha) {
      /* let the checkmark or radio circle already take up space, else when checking everyhing collapses */
      @apply h-2/3 w-2/3;
    }
    &[role='radio']::part(ch) {
      @apply flex h-5 w-5 shrink-0 items-center justify-center rounded-full shadow-sm ring-2 ring-inset ring-gray-300;
    }
    &[role='radio'][aria-checked='true']::part(cha) {
      @apply rounded-full bg-primary-500;
    }
    &[role='checkbox']::part(ch) {
      @apply flex h-5 w-5 shrink-0 items-center justify-center rounded shadow-sm ring-2 ring-inset ring-gray-300;
    }
    &[role='checkbox'][aria-checked='true']::part(cha) {
      @apply bg-primary-500;
      -webkit-mask: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' fill='currentColor' width='100%' height='100%' viewBox='0 0 24 24'%3E%3Cpath d='M20.285 2l-11.285 11.567-5.286-5.011-3.714 3.716 9 8.728 15-15.285z'/%3E%3C/svg%3E");
    }
  }
  /* - text    : text-entry and extended-text */
  .text {
    @apply base cursor-text rounded-sm border-0 p-3 text-gray-600 shadow-inner;
  }

  /* - spot    : gap-select-point, graphic-order, qti-graphic-associate */
  .spot {
    @apply base h-full w-full bg-transparent p-0 ring-2 ring-gray-900;
  }

  .point {
    @apply base h-6 w-6 bg-transparent p-0 ring-2 ring-gray-900;
  }

  /* - drag    : gap-match, match, order, associate */
  .drag {
    @apply base cursor-grab bg-white py-1 pl-4 pr-2 font-semibold shadow-sm;
    /* @apply select-none; */
    background-image: radial-gradient(
      circle at center,
      rgba(0, 0, 0, 0.1) 0,
      rgb(0, 0, 0, 0.1) 2px,
      white 2px,
      white 100%
    );
    background-repeat: repeat-y;
    background-position: left 2px;
    background-size: 14px 8px;
  }

  /* - drop    : gap-match, match, order, associate */
  .drop {
    @apply base ring-outset relative bg-gray-50 shadow-inner ring-2 ring-gray-300 ring-offset-1;
  }

  .order {
    @apply h-6 w-6 rounded-full bg-primary-500 text-white;
  }
  /** End components */
}

@layer utilities {
  /* hover, every element has a hover, except disabled( which overrides hover */
  .hov {
    @apply bg-gray-50;
  }
  .hov-ring {
    @apply ring-primary-500;
  }
  /* focus, every element has a focus, even drop ( for keyboard access, eventually ) */
  .foc {
    @apply outline-2 outline-primary-300;
  }
  /* active, choice active class and, used on checkbox and radio where not .qti-input-choice-hidden */
  .act {
    @apply ring-2 ring-primary-400;
  }
  /* readonly, overwrites cursor, hover background, outline for focus and ring */
  .rdo {
    @apply cursor-pointer bg-white shadow-none outline-0 ring-0;
  }
  /* disabled, make gray state */
  .dis {
    @apply cursor-not-allowed bg-gray-100 text-gray-500 shadow-none ring-gray-400;
    /* drag in the drop, else the drag in the drop was black and not gray */
    & > * {
      @apply !text-gray-500;
    }
  }
}

@layer utilities {
  /** Modifiers */

  /* PK: There was no utility to reset the ring-inset
  We use this for the drop container, where we want the ring to have an outset
  So the drag, with ring-inset, snugs neatly into the drop with ring-'outset' 
  by unsetting the ring-inset */
  .ring-outset {
    --tw-ring-inset: unset;
  }

  /* intext for elements which has to appear in text
  - intext    : gap-match > qti-gap, hottext > hottext, select
  */
  .button-in {
    @apply inline-flex;

    /* @apply inline-flex gap-1 px-1 py-0 align-bottom font-normal ring-offset-0; */
  }
  .check-in {
    @apply inline-flex;

    /* @apply inline-flex gap-1 px-1 py-0 align-bottom font-normal ring-offset-0; */
  }

  /* small for elements which has to appear in text
  - intext    : gap-match > qti-gap, hottext > hottext, select
*/
  .button-sm {
    /* @apply gap-0.5 rounded !px-1 !py-0 outline-offset-0 ring-1; */
  }
  .check-sm {
    @apply gap-0.5 outline-offset-0 ring-1;
    /* @apply gap-0.5 rounded !px-1 !py-0 outline-offset-0 ring-1; */
    &::part(ch) {
      @apply !h-4 !w-4 !ring-1;
    }
  }
}
