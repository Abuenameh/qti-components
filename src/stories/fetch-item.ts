import { qtiTransformItem, qtiTransformTest } from 'src/lib/qti-transformers';

export const fetchItem = async (packageUri: string, index: number): Promise<any> => {
  packageUri = packageUri.endsWith('/') ? packageUri : packageUri + '/';
  const assessmentTestHref = `assessment.xml`;
  const itemLocation = `${packageUri}${assessmentTestHref.substring(0, assessmentTestHref.lastIndexOf('/'))}/`;

  const itemsFromTest = await qtiTransformTest()
    .load(packageUri + assessmentTestHref)
    .then(api => api.items());

  const itemHTML = await qtiTransformItem()
    .load(itemLocation + itemsFromTest[index].href)
    .then(api => api.path(itemLocation).html());

  return { itemXML: itemHTML, items: itemsFromTest };
};
