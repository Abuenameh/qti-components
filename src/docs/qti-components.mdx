import { Canvas, Meta, Story } from '@storybook/blocks';
import './documentation.tailwind.css';

<Meta title="QTI Components" />

# QTI Components

Using the standard for non-standard tests

## Promise

Renders basic QTI3 XML's.
QTI components are an implementation of the QTI standard in webcomponents

## Installation

```
npm install @citolab/qti-components@latest
```

```js
import '@citolab/qti-components';
```

https://jsfiddle.net/mrklein/s97Ld0gn/20/
https://codepen.io/trickyjs/pen/xxyyWLG

import * as StylesheetStories from '../lib/qti-components/qti-interaction/qti-choice-interaction/qti-choice-interaction.stories';

### Choice

<Canvas of={StylesheetStories.Interaction} />

<iframe
  width="100%"
  height="300"
  src="//jsfiddle.net/mrklein/s97Ld0gn/embedded/result,html/dark/"
  allowfullscreen="allowfullscreen"
  allowpaymentrequest
  frameborder="0"
></iframe>

<iframe
  width="100%"
  height="300"
  src="//stackblitz.com/edit/qti-player?embed=1&file=src%2FPlayer.tsx&hideExplorer=1"
  allowfullscreen="allowfullscreen"
  allowpaymentrequest
  frameborder="0"
></iframe>

## Features

## Principles

- Offline first
- Composable, leave out what you do not need
- Zero configuration
- Zero option flags
- Composable, just like lego bricks you can you make your own player, review or builder
- Serverless, rendering and scoring can be done on the frontend
- Responsive, support for different screen resolutions
- Extensible with customized qti-interactions ( see qti-components-custom )
- No clever code (ie hacks!)
- Embraces different platforms ( touch/mouse ) and their way of input
- Pluggable responseprocessing ( see qti-components-responseprocessing )
- Framework agnostic, use with Vue, React, Angular or plain vanilla
- Lightweight, core bundle is just 17kb not minified
- Few dependencies, only one dependency on @Lit
- Stateless, event based, no state is managed
- Expose relevant components to light dom for easy styling
- Examples in plain vanilla javascript

# How to use

- qti-item
- qti-components
- qti-transform
