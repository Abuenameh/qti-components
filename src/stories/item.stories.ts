import '@citolab/qti-components/qti-components';

import { action } from '@storybook/addon-actions';
import type { Meta, StoryObj } from '@storybook/web-components';
import { html } from 'lit';
import { unsafeHTML } from 'lit/directives/unsafe-html.js';
import packages from '../assets/packages.json';
import { fetchItem } from './fetch-item';

const meta: Meta = {
  title: 'Examples',
  argTypes: {
    qtipkg: {
      options: packages.packages,
      control: { type: 'radio' }
    },
    disabled: { control: { type: 'boolean' } }
  },
  args: {
    serverLocation: 'http://localhost:6006/api',
    qtipkg: 'examples',
    itemIndex: 0
  }
};

export default meta;
type Story = StoryObj;

let item;

export const Examples: Story = {
  render: ({ disabled, view }, { argTypes, loaded: { xml } }) => {
    const onInteractionChangedAction = action('on-interaction-changed');
    const onOutcomeChangedAction = action('qti-outcome-changed');
    const onItemFirstUpdated = ({ detail: qtiAssessmentItem }) => {
      item = qtiAssessmentItem;
      action('qti-item-first-updated');
    };

    item && (item.disabled = disabled);
    item && (item.view = view);

    return html`
      <div
        @qti-interaction-changed=${onInteractionChangedAction}
        @qti-outcome-changed=${onOutcomeChangedAction}
        @qti-item-first-updated=${onItemFirstUpdated}
      >
        ${unsafeHTML(xml.itemXML)}
      </div>
      <button style="background:lightgray; padding:1rem" @click=${() => item?.processResponse()}>Submit</button>
    `;
  },
  loaders: [async ({ args }) => ({ xml: await fetchItem(`${args.serverLocation}/${args.qtipkg}`, args.itemIndex) })]
};
