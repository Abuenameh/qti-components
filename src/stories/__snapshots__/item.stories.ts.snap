// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`items Items smoke-test 1`] = `
<div id="root-inner">
  <div>
    <qti-assessment-item xmlns="http://www.w3.org/1999/xhtml"
                         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                         xsi:schemalocation="http://www.imsglobal.org/xsd/imsqtiasi_v3p0 https://purl.imsglobal.org/spec/qti/v3p0/schema/xsd/imsqti_asiv3p0_v1p0.xsd"
                         title="32llxx"
                         identifier="ITM-32llxx"
                         time-dependent="false"
                         label="32llxx"
    >
      <qti-stylesheet href="http://localhost:6006/api/biologie/depitems/..%2Fcss%2Fcito_itemstyle.css"
                      type="text/css"
      >
      </qti-stylesheet>
      <qti-stylesheet href="http://localhost:6006/api/biologie/depitems/..%2Fcss%2Fcito_userstyle.css"
                      type="text/css"
      >
      </qti-stylesheet>
      <qti-stylesheet href="http://localhost:6006/api/biologie/depitems/..%2Fcss%2Fcito_generated.css"
                      type="text/css"
      >
      </qti-stylesheet>
      <qti-item-body class="defaultBody">
        <div class="content">
          <div class="qti-layout-row">
            <div class="qti-layout-col12">
              <p>
                <strong>
                  Let op!
                </strong>
              </p>
              <p>
                Dit centraal examen bevat één of meer onvolkomenheden.
              </p>
              <table class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_1">
                <colgroup>
                  <col class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_2">
                  <col class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_3">
                </colgroup>
                <tbody>
                  <tr>
                    <td class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_4">
                      <p class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_5">
                        <strong>
                          Vraagnummer
                        </strong>
                      </p>
                    </td>
                    <td class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_6">
                      <p class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_7">
                        <strong>
                          Aard van de onvolkomenheid
                        </strong>
                      </p>
                    </td>
                  </tr>
                  <tr>
                    <td class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_8">
                      <p class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_9">
                        18
                      </p>
                    </td>
                    <td class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_10">
                      <p class="cito_genclass_Infoscherm_onvolkomenheden_2022_-_examenvariant_1_11">
                        Uitwendig behoort tot KB deel van de syllabus. Besloten is om hiervoor te compenseren.
                      </p>
                    </td>
                  </tr>
                </tbody>
              </table>
              <p>
              </p>
            </div>
          </div>
        </div>
      </qti-item-body>
      <link rel="stylesheet"
            type="text/css"
            media="screen"
            href="http://localhost:6006/api/biologie/depitems/..%2Fcss%2Fcito_itemstyle.css"
      >
      <style media="screen">
      </style>
      <link rel="stylesheet"
            type="text/css"
            media="screen"
            href="http://localhost:6006/api/biologie/depitems/..%2Fcss%2Fcito_userstyle.css"
      >
      <style media="screen">
      </style>
      <link rel="stylesheet"
            type="text/css"
            media="screen"
            href="http://localhost:6006/api/biologie/depitems/..%2Fcss%2Fcito_generated.css"
      >
      <style media="screen">
      </style>
    </qti-assessment-item>
  </div>
</div>
`;
